### ISSUE TITLE
Codeaction to import missing function

### ISSUE THREAD

**Issue by andys8** [nwolverson/purescript-language-server#85]
Currently it is possible to import functions/modules with completions.

The import functionality is there, one has to go in input mode, and trigger completion:
![image](https://user-images.githubusercontent.com/13085980/90962727-6b13e900-e4b2-11ea-91e8-4c6dd9de58be.png)

It would be nice if it would also to have a code action to resolve issues like this :)

![image](https://user-images.githubusercontent.com/13085980/90962706-4c155700-e4b2-11ea-9714-87c6b1071763.png)

**Comment by andys8** [comment]
Oh, wow, that's interesting. The good news: vscode is working very well! The bad news: It doesn't work with [Vim/Coc](https://github.com/neoclide/coc.nvim) :(

I looked into messages and I think it might have to do with expectations regarding ranges in the code.

### Example

![image](https://user-images.githubusercontent.com/13085980/90985631-73852600-e57d-11ea-8a6c-1d8afe3768cf.png)
https://github.com/jonasbuntinx/purescript-react-realworld/blob/master/src/Main.purs

### vscode

```
[Trace - 7:41:08 PM] Sending request 'textDocument/codeAction - (17)'.
Params: {
    "textDocument": {
        "uri": "file:///home/andreas/dev/repository/purescript-react-realworld/src/Main.purs"
    },
    "range": {
        "start": {
            "line": 21,
            "character": 20
        },
        "end": {
            "line": 21,
            "character": 20
        }
    },
    "context": {
        "diagnostics": [
            {
                "range": {
                    "start": {
                        "line": 21,
                        "character": 15
                    },
                    "end": {
                        "line": 21,
                        "character": 20
                    }
                },
                "message": "  Unknown type Unit1\n",
                "severity": 1,
                "code": "UnknownName",
                "source": "PureScript"
            }
        ]
    }
}


[Trace - 7:41:08 PM] Received response 'textDocument/codeAction - (17)' in 12ms.
Result: [
    {
        "title": "Fix typo/add import",
        "command": "purescript.fixTypo",
        "arguments": [
            "file:///home/andreas/dev/repository/purescript-react-realworld/src/Main.purs",
            21,
            15
        ]
    }
]
```

### Vim/CoC

#### coc-codeaction-line

https://github.com/neoclide/coc.nvim/blob/15f1cb039b65e9a0655411b8fec51fe5f94ab607/doc/coc.txt#L1002

```
[Trace - 7:39:26 PM] Sending request 'textDocument/codeAction - (3)'.
Params: {
    "textDocument": {
        "uri": "file:///home/andreas/dev/repository/purescript-react-realworld/src/Main.purs"
    },
    "range": {
        "start": {
            "line": 21,
            "character": 0
        },
        "end": {
            "line": 21,
            "character": 20
        }
    },
    "context": {
        "diagnostics": [
            {
                "range": {
                    "start": {
                        "line": 21,
                        "character": 15
                    },
                    "end": {
                        "line": 21,
                        "character": 20
                    }
                },
                "severity": 1,
                "code": "UnknownName",
                "source": "PureScript",
                "message": "  Unknown type Unit1\n"
            }
        ]
    }
}


[Trace - 7:39:26 PM] Received response 'textDocument/codeAction - (3)' in 12ms.
Result: []
```


#### coc-codeaction

https://github.com/neoclide/coc.nvim/blob/15f1cb039b65e9a0655411b8fec51fe5f94ab607/doc/coc.txt#L998

```
[Trace - 7:46:08 PM] Sending request 'textDocument/codeAction - (2)'.
Params: {
    "textDocument": {
        "uri": "file:///home/andreas/dev/repository/purescript-react-realworld/src/Main.purs"
    },
    "range": {
        "start": {
            "line": 0,
            "character": 0
        },
        "end": {
            "line": 44,
            "character": 0
        }
    },
    "context": {
        "diagnostics": [
            {
                "range": {
                    "start": {
                        "line": 21,
                        "character": 15
                    },
                    "end": {
                        "line": 21,
                        "character": 20
                    }
                },
                "severity": 1,
                "code": "UnknownName",
                "source": "PureScript",
                "message": "  Unknown type Unit1\n"
            }
        ]
    }
}


[Trace - 7:46:08 PM] Received response 'textDocument/codeAction - (2)' in 17ms.
Result: []
```

## Investigation

By changing [`contains`](https://github.com/andys8/purescript-language-server/blob/e37acd8a4c16760a2d808570f74d6501b71eb332/src/LanguageServer/IdePurescript/CodeActions.purs#L84) I could see code actions. But applying them didn't work. But not sure what's happening in [`fixTypo`](https://github.com/andys8/purescript-language-server/blob/e37acd8a4c16760a2d808570f74d6501b71eb332/src/LanguageServer/IdePurescript/Assist.purs#L99-L118)

**Comment by nwolverson** [comment]
Without looking at the code again, I suspect the filter for code actions should be an intersection of the request and diagnostic ranges

**Comment by andys8** [comment]
Filtering is one thing. But also Codeaction application was not working. I'm eager to come back to that issue because it would be a large improvement for language server users.

**Comment by nwolverson** [comment]
The code action has client-side support in vscode actually, required to support multiple options for import - the implementation uses both client and server-side command implementations (though I think if there's only 1 possible option it will otherwise work). Not sure how possible this is with other LSP implementations.

I'd really recommend just triggering manually completion/import :)

**Comment by andys8** [comment]
Hmm. I'm not sure if I understood correctly what you wrote.

From the lap client side (CoC/Vim) fixing diagnostics by applying a code actions to import the missing import (including selecting it) is working with other language servers like Haskell, Elm, and TypeScript. I have to double check if I have that correct in mind.

With PureScript it is working with vscode. So my guess is there will be a way to make it work the same way with Vim or implement it in a way that works with all clients.

**Comment by andys8** [comment]
# Other Language Servers

## Haskell

> ghcide version: 0.1.0 (GHC: 8.8.3) (PATH: /home/andreas/.local/bin/ghcide) (GIT hash: 1f90cc2b70c1c98277e771566e2b9e9ae9c0837f)


![image](https://user-images.githubusercontent.com/13085980/92324812-594f4b80-f045-11ea-9f2a-22129148884a.png)

**Comment by nwolverson** [comment]
Neither of those examples are quite the same as what we have just now. There is 1 code action, which triggers a list of replacement options to select, this is enabled by client-side code, allows a nice dropdown with types etc.

I can see the examples you give are showing multiple code-actions, one for each possible import etc. This is definitely something we could (should) do instead, though the UI will be rather restricted compared to what is present now in vscode where we can present a dropdown with all module and type information

**Comment by andys8** [comment]
You're right. Now the difference is much clearer to me.

## Vim

![image](https://user-images.githubusercontent.com/13085980/92324950-d7f8b880-f046-11ea-91ef-3cd74a7eba8b.png)

There is currently *no* action at all, but that's probably a different problem (the filter / check whether the range is matching).

## VS Code

![image](https://user-images.githubusercontent.com/13085980/92324961-fe1e5880-f046-11ea-945d-47408ce87f54.png)

![image](https://user-images.githubusercontent.com/13085980/92324977-11312880-f047-11ea-8bb0-8086710d0745.png)

**Comment by nwolverson** [comment]
Implemented codeaction per suggestion in ef1247befe443933ce56a3a12abc7cb745ddf794

The filtering is a little off just now, at least in VSCode it's matching at and after the end of the symbol

**Comment by andys8** [comment]
Wow, that's awesome! 😀

I think I found the bug which made actions not showing up at all for me. Maybe this helps with the vscode issue too.

https://github.com/nwolverson/purescript-language-server/pull/93

**Comment by andys8** [comment]
With ef1247b and #93 merged, this issue should solved!

**Comment by nwolverson** [comment]
Seems like the add import part is not working still (0.14.2)

**Comment by nwolverson** [comment]
Aha, it was fixed and I broke it, be55250327ca1c881235b6655a2cc3b5481c3d83 should do it...

**Comment by andys8** [comment]
I'll give it a try
